gmt-4.5.14/src/misc/kml2gmt.1
.TH KML2GMT 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
kml2gmt \- Extract GMT table data from Google Earth KML files 
.SH SYNOPSIS
\fBkml2gmt\fP [ \fIinfile\fP ] [ \fB\-V\fP ] [ \fB\-Z\fP ] [ \fB\-:\fP[\fBi\fP|\fBo\fP] ] [ \fB\-bo\fP[\fBs\fP|\fBS\fP|\fBd\fP|\fBD\fP[\fIncol\fP]|\fBc\fP[\fIvar1\fP\fB/\fP\fI...\fP]] ] 
.SH DESCRIPTION
\fBkml2gmt\fP reads a  Google Earth KML file and outputs a GMT table file.
Only KML files that contain points, lines, or polygons can be processed.
This is a bare-bones operation that aims to extract coordinates and possibly the
name and description tags of each feature.  The main use intended is to capture coordinates modified
in Google Earth and then reinsert the modified data into the original
GMT data file.  For a more complete reformatting, consider using \fBogr2ogr \-f\fP
"GMT" somefile.gmt somefile.kml.
.TP
\fIinfile\fP
Name of the KML file to work on.  If not given, standard input is read.
.SH OPTIONS
No space between the option flag and the associated arguments.
.TP
\fB\-Z\fP
Output the altitude coordinates as GMT z coordinates [Default will output just
longitude and latitude].
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.TP
\fB\-:\fP
Toggles between (longitude,latitude) and (latitude,longitude) input and/or output.  [Default is (longitude,latitude)].
Append \fBi\fP to select input only or \fBo\fP to select output only.  [Default affects both].
.TP
\fB\-bo\fP
Selects binary output.
Append \fBs\fP for single precision [Default is \fBd\fP (double)].
Uppercase \fBS\fP or \fBD\fP will force byte-swapping.
Optionally, append \fIncol\fP, the number of desired columns in your binary output file.
.SH EXAMPLES
.SH "SEE ALSO"
.IR gmtdefaults (1),
.IR GMT (1),
.IR img2google (1),
.IR ps2raster (1)
.IR gmt2kml (1)


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/misc/kml2gmt.c
/*--------------------------------------------------------------------
 *	$Id: kml2gmt.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *	Copyright (c) 2009-2015 by P. Wessel
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; version 2 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/*
 * kml2gmt is a reformatter that takes KML files and extracts GMT tables;
 * it is the opposite of gmt2kml.
 *
 * Author:	Paul Wessel
 * Date:	19-MAY-2009
 * Version:	1.0
 */
 
#include "gmt.h"

#define POINT			0
#define LINE			1
#define POLYGON			2

int main (int argc, char **argv)
{
	GMT_LONG i, n_files = 0, ncol = 2, start, fmode = POINT;
	GMT_LONG scan = TRUE, first = TRUE;

	GMT_LONG error = FALSE;
	
	char line[BUFSIZ], header[BUFSIZ], name[BUFSIZ], description[BUFSIZ], *file = NULL;

	double out[3];

	FILE *fp = NULL;

	GMT_LONG ascii_output_one (double x, GMT_LONG col);

	argc = (int)GMT_begin (argc, argv);

	GMT_io.in_col_type[GMT_X] = GMT_io.out_col_type[GMT_X] = GMT_IS_LON;
	GMT_io.in_col_type[GMT_Y] = GMT_io.out_col_type[GMT_Y] = GMT_IS_LAT;
	memset ((void *)header, 0, BUFSIZ);
	memset ((void *)name, 0, BUFSIZ);
	memset ((void *)description, 0, BUFSIZ);
	
	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {

				/* Common parameters */

				case 'V':
				case 'b':
				case 'm':
				case ':':
				case '\0':
					error += GMT_parse_common_options (argv[i], NULL, NULL, NULL, NULL);
					break;

				/* Supplemental parameters */

				case 'Z':
					ncol = 3;
					break;
				default:
					error = TRUE;
					GMT_default_error (argv[i][1]);
					break;
			}
		}
		else {
			file = argv[i];
			n_files++;
		}
	}

	if (error || GMT_give_synopsis_and_exit) {
		fprintf (stderr, "kml2gmt %s - Extract GMT data from a Google Earth KML file\n\n", GMT_VERSION);
		fprintf (stderr, "usage: kml2gmt [<infile>] [-V] [%s] [%s] [%s] > GMTdata.txt\n", GMT_m_OPT, GMT_t_OPT, GMT_bo_OPT);

		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);

		fprintf (stderr, "\tinfile is the Google Earth KML file.\n");
		fprintf (stderr, "\t  If no file(s) is given, standard input is read.\n");
		fprintf (stderr, "\n\tOPTIONS:\n");
		fprintf (stderr, "-Z Output the z-column from the KML file [Only lon,lat is output]\n");
		GMT_explain_option ('V');
		GMT_explain_option ('m');
		GMT_explain_option (':');
		GMT_explain_option ('o');
		GMT_explain_option ('n');
		GMT_explain_option ('.');

		exit (EXIT_FAILURE);
	}

	if (n_files > 1) {
		fprintf (stderr, "%s: GMT SYNTAX ERROR.  Only one file can be processed at the time\n", GMT_program);
		error++;
	}
	if (GMT_io.binary[GMT_IN] && GMT_io.io_header[GMT_IN]) {
		fprintf (stderr, "%s: GMT SYNTAX ERROR.  Binary input data cannot have header -H\n", GMT_program);
		error++;
	}
        if (GMT_io.binary[GMT_IN] && GMT_io.ncol[GMT_IN] == 0) GMT_io.ncol[GMT_IN] = 2;
	GMT_io.multi_segments[GMT_OUT] = TRUE;	/* So we can write segments */
	if (error) exit (EXIT_FAILURE);

	if (GMT_io.binary[GMT_IN] && gmtdefs.verbose) {
		char *type[2] = {"double", "single"};
		fprintf (stderr, "%s: Expects %ld-column %s-precision binary data\n", GMT_program, GMT_io.ncol[GMT_IN], type[GMT_io.single_precision[GMT_IN]]);
	}
#ifdef SET_IO_MODE
	GMT_setmode (GMT_OUT);
#endif
	/* Now we are ready to take on some input values */

	strcpy (gmtdefs.d_format, "%.12g");	/* Get enough decimals */
	
	if (n_files == 0) {	/* Just read standard input */
		fp = stdin;
		if (gmtdefs.verbose) fprintf (stderr, "%s: Reading from standard input\n", GMT_program);
#ifdef SET_IO_MODE
		GMT_setmode (GMT_IN);
#endif
		printf ("# %s: KML read from standard input\n", GMT_program);
	}
	else if ((fp = fopen (file, GMT_io.r_mode)) == NULL) {
		fprintf (stderr, "%s: Cannot open file %s\n", GMT_program, file);
		exit (EXIT_FAILURE);
	}
	else {
		if (gmtdefs.verbose) fprintf (stderr, "%s: Processing %s\n", GMT_program, file);
		printf ("# %s: KML read from %s\n", GMT_program, file);
	}

	while (fgets (line, BUFSIZ, fp)) {
		if (strstr (line, "<Placemark")) scan = TRUE;
		if (strstr (line, "</Placemark")) scan = FALSE;
		if (!scan) continue;
		if (strstr (line, "<Point")) fmode = POINT;
		if (strstr (line, "<LineString")) fmode = LINE;
		if (strstr (line, "<Polygon")) fmode = POLYGON;
		if (strstr (line, "<name>")) {
			for (i = 0; i < (GMT_LONG)strlen (line) && line[i] != '>'; i++);	/* Find end of <name> */
			start = i + 1;
			for (i = start; i < (GMT_LONG)strlen (line) && line[i] != '<'; i++);	/* Find start of </name> */
			line[i] = '\0';
			strcpy (name, &line[start]);
			GMT_chop (name);
			if (first) printf ("# %s\n", &line[start]);
			first = FALSE;
		}
		if (strstr (line, "<description>")) {
			for (i = 0; i < (GMT_LONG)strlen (line) && line[i] != '>'; i++);	/* Find end of <description> */
			start = i + 1;
			for (i = start; i < (GMT_LONG)strlen (line) && line[i] != '<'; i++);	/* Find start of </description> */
			line[i] = '\0';
			strcpy (description, &line[start]);
			GMT_chop (description);
			if (first) printf ("# %s\n", &line[start]);
			first = FALSE;
		}
		if (name[0] || description[0]) {
			sprintf (GMT_io.segment_header, "%c", GMT_io.EOF_flag[GMT_OUT]);
			if (name[0]) { strcat (GMT_io.segment_header, " -L\""); strcat (GMT_io.segment_header, name); strcat (GMT_io.segment_header, "\""); }
			if (description[0]) { strcat (GMT_io.segment_header, " -D\""); strcat (GMT_io.segment_header, description); strcat (GMT_io.segment_header, "\""); }
			strcat (GMT_io.segment_header, "\n");
		}
		
		if (!strstr (line, "<coordinates>")) continue;
		/* We get here when the line says coordinates */
		if (fmode == POINT) {	/* Process the single point */
			for (i = 0; i < (GMT_LONG)strlen (line) && line[i] != '>'; i++);		/* Find end of <coordinates> */
			sscanf (&line[i+1], "%lg,%lg,%lg", &out[GMT_X], &out[GMT_Y], &out[GMT_Z]);
			if (gmtdefs.xy_toggle[GMT_OUT]) d_swap (out[GMT_X], out[GMT_Y]);		/* Output lat/lon instead of lon/lat */
			ascii_output_one (out[GMT_X], GMT_X);	printf ("%s", gmtdefs.field_delimiter);
			ascii_output_one (out[GMT_Y], GMT_Y);	
			if (ncol == 3) { printf ("%s", gmtdefs.field_delimiter); ascii_output_one (out[GMT_Z], GMT_Z);}
			printf ("\n");
		}
		else {
			if (GMT_io.segment_header[0]) printf ("%s", GMT_io.segment_header);
			name[0] = description[0] = 0;
			while (fscanf (fp, "%lg,%lg,%lg", &out[GMT_X], &out[GMT_Y], &out[GMT_Z])) {
				if (gmtdefs.xy_toggle[GMT_OUT]) d_swap (out[GMT_X], out[GMT_Y]);		/* Output lat/lon instead of lon/lat */
				ascii_output_one (out[GMT_X], GMT_X);	printf ("%s", gmtdefs.field_delimiter);
				ascii_output_one (out[GMT_Y], GMT_Y);	
				if (ncol == 3) { printf ("%s", gmtdefs.field_delimiter); ascii_output_one (out[GMT_Z], GMT_Z);}
				printf ("\n");
			}
		}
	}
	
	if (fp != stdin) fclose (fp);
	
	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}

GMT_LONG ascii_output_one (double x, GMT_LONG col)
{	/* Used instead of GMT_asfii_output_one since Windoze has trouble with GMT_stdout and stdout mixing */
	char text[GMT_LONG_TEXT];

	GMT_ascii_format_one (text, x, GMT_io.out_col_type[col]);
	return (printf ("%s", text));
}
